### ISSUE TITLE
As a user I want to be able to restrict IP families to some pods

### ISSUE THREAD

**Issue by chandanD4** [kubernetes/kubernetes#109814]
### What would you like to be added?

In dual stack cluster ,  all the pods have both Ipv6 and Ipv4 interfaces. However there would be scenarios where user does not to enable dual stack for all the pods . So basically the ask is to restrict IP families to certain POD.  We should be able to specify certain pods to operate in single stack only i.e. either IPv4 or IPv6 only.  The address family of a Pod defaults to the address family of the first cluster-cidr IP range (configured via the --cluster-cidr flag to the kube-apiserver).. 

One of possible way to indicate in the pod spec is shown below

apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
ipFamilyPolicy: [PreferDualStack , SingleStack, DualStack]
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80

### Why is this needed?

This would prevent legacy apps /Pod/ single stack pod to operate securely without unintended opening holes on additional interfaces. This would prevent developer unintended support for secondary IP families and thereby helping secure deployment.

**Comment by aojea** [comment]
/sig network

I've received similar questions recently and this is trickier than with services because PodIPs are really out of control of the Cluster, they belong to the CNI, and the CNI may not use ClusterCIDR for the Pods IP assignment.

At high level the flow to create a pod and report the IPs is:
 kubelet -> cri -> cni -> (assign IPs) -> cri -> kubelet -> (update pod.Status.PodIPs to the apiserver)

I wonder if this falls more into the network policy realm @danwinship @uablrek @thockin ....

**Comment by uablrek** [comment]
> I wonder if this falls more into the network policy realm

I totally agree on that. A general `-j DROP` in `iptables` or `ip6tables` (respectively) should cover it.

**Comment by aojea** [comment]
> You'd really need ipFamilies too.

it tooks more than 2 years to settle on that and was a lot of work, do you think we can do this at this point for Pods?
it looks that train has left the station alread


> > I wonder if this falls more into the network policy realm
> 
> I don't think so. We should not give pods pod IPs that don't actually work. If nothing else, it will be inefficient, as pods will sometimes try to make IPv6 connections that are doomed to fail, before falling back to IPv4 (or vice versa).
> 
> >

let's assume that the user don't have control over the cni

IF I have cluster with ipv4 pod cidr 10.0.0.0/8 and an ipv6 pod cidr 2001:db8::/64.
AND my pods ALWAYS get dual-stack addresses
BUT I want pod X to work only in IPv4
CAN I use a network policy to say "Allow Only IPv4 communications" ?

@danwinship  is not possible to use the network policy?
```
        - ipBlock:
            cidr: 10.0.0.0/24
```

**Comment by sftim** [comment]
The workload might not do [fast fallback](https://en.wikipedia.org/wiki/Happy_Eyeballs) to IPv4: if it sees that the target has an IPv6 address, it might try reaching it and only fall back to IPv4 after (say) 180 seconds.

Not ideal. I think there _is_ a difference between not presenting an IPv6 address to a Pod, and presenting it but dropping / rejecting all its packets.

PS. Using rejection would allow faster fallback compared to dropping packets, but:
- NetworkPolicy doesn't let you pick between drop and reject
- sending back a rejection datagram is quite different from the brief to restrict IP address family. In the rejection case, you're actively sending with the unwanted IP family.

**Comment by danwinship** [comment]
> > You'd really need ipFamilies too.
> 
> it tooks more than 2 years to settle on that and was a lot of work, do you think we can do this at this point for Pods?

No no no. It took us a year to even decide that we _wanted_ dual-stack services, and then it took us half a year to come up with a suitable semantics for dual-stack services, and half a year to add the implementation of dual-stack services to the existing implementation of single-stack services.

We already have dual-stack pods, and core Kubernetes is perfectly capable of dealing with a cluster with a mix of single-stack and dual-stack pods. We just don't have an API to explicitly request that a particular pod should end up single-stack or dual-stack. But that's trivial.

> IF I have cluster with ipv4 pod cidr 10.0.0.0/8 and an ipv6 pod cidr 2001:db8::/64.
> AND my pods ALWAYS get dual-stack addresses
> BUT I want pod X to work only in IPv4

Then just run the software in such a way that it won't accept IPv6 connections? (eg, `./myserver --bind-addr ${NODE_IP}`)

> CAN I use a network policy to say "Allow Only IPv4 communications" ?

Not usefully. You can say "allow connections from all IPv4 addresses" or "allow connections from all IPv4 pod IPs" or "allow connections from 10.0.1.2 and 10.2.3.4". But as soon as you want to use a `podSelector` or `namespaceSelector` then the game is up.

**Comment by thockin** [comment]
> The API may be trivial, but I have doubts about the implementation, ... the CNI don't have to use the API and keep assigning the families they want

This.

Our choices, as I see it:

1) Do nothing in core, it's a network plugin problem
2) Soft-standardize one or more annotations and recommend net plugins respect them
3) Add a pod API to specify this and recommend net plugins respect them

No matter what we can't REQUIRE plugins to do anything, so it will be optional for the foreseeable future.

Option 3 starts to feel a whole lot like multi-network...

**Comment by uablrek** [comment]
> A use case I heard recently is "IPv6-mostly". A customer is already using almost all of 10.0.0.0/8 for other internal purposes, and doesn't want to have to carve out large chunks of it to support lots of pods in their "dual-stack" cluster when most of the pods will never need to use IPv4. So they want a cluster where most of the pods are single-stack IPv6, but a few are dual-stack IPv6-primary.

IMHO this is *the* use-case. I can't really imagine that the presence of an IPv6 address on eth0 opens security holes, even for legacy applications. But when IPv6 is used one don't want the constraints imposed by the IPv4 address range.

My vote is on;
> Do nothing in core, it's a network plugin problem

Network plugins will respond to this if many users want it, and possibly in ways we can't see yet. To "Soft-standardize one or more annotations" I think _must_ be avoided. At least until we see what users want (compare with the [bandwidth annotation](https://github.com/kubernetes/enhancements/pull/2808)).

> Option 3 starts to feel a whole lot like multi-network...

With one important difference; multi-network is more to extend networking in PODs, this issue is to restrict it. There is a bunch of "what if's", e.g a dual-stack service with a selector that points out a mix of ipv4,ipv6 and dual-stack PODs.

**Comment by thockin** [comment]
>> Option 3 starts to feel a whole lot like multi-network...
> 
> With one important difference; multi-network is more to extend networking in PODs, this issue is to restrict it. There is a bunch of "what if's", e.g a dual-stack service with a selector that points out a mix of ipv4,ipv6 and dual-stack PODs.

If you model the "pod network" as 2 networks (v4, v6) and allow pods to specificy which one(s) they want, maybe it can be a fit.  Maybe.  Very speculative since we don't have multi-network as a built-in concept.

**Comment by danwinship** [comment]
I guess just adding `ipFamilies`/`ipFamilyPolicy` to Pod wouldn't really solve the "IPv6-mostly" use case, because `PreferDualStack` would still presumably be the default, so someone who wanted "IPv6-mostly" would have to manually override `ipFamilyPolicy` on almost all of their pods. (Well, unless we also added an apiserver option to set the default...)

One possibility: suggest that plugins should default to the same IP-family-ness for pods that kubelet does for the pod's node. ie, If kubelet decides that the node has only an IPv4 IP, then plugins should create single-stack IPv4 pods. If kubelet decides that the node has only an IPv6 IP, then they should create single-stack IPv6 pods. If kubelet decides that the node has dual-stack IPs, then they should create dual-stack pods. (Kubelet already sorts `pod.podIPs` to match `pod.hostIPs`, so this would just be extending that to say that plugins should try to match `pod.hostIPs` to begin with.)

(Of course, CNI doesn't currently provide any way for the plugin to know `pod.hostIPs` but whatever.)

This kind of makes sense anyway, since if a node is single-stack, then dual-stack pods on that node will likely not be 100%-functional anyway (eg, with most plugins, pod-to-cluster-external traffic wouldn't work for the missing IP family).

So then a cluster could have a mix of IPv6-only nodes and dual-stack nodes, and, eg, you could taint the dual-stack nodes, and then pods would only get dual-stack IPs if they were explicitly marked as "tolerating" dual-stack. (Or you could use various other means of correctly scheduling pods to nodes.)

**Comment by uablrek** [comment]
I made an experiment with a mix of dual-stack and ipv4/ipv6-only PODs. I use a very simple `bridge` cni-plugin and edit the `/etc/cni/net.d/10-bridge.conf` between creation of `Deployments`. Just altering the `ipam` section. K8s has no problem with this mix of PODs, neither does `cri-o` which is my CRI-plugin.
```
vm-002 ~ # alias pods
pods='kubectl get pods -o '"'"'custom-columns=NAME:metadata.name,NODE:spec.nodeName,IPs:status.podIPs[*].ip'"'"
vm-002 ~ # pods
NAME                    NODE     IPs
dual-59794d75c6-22926   vm-001   11.0.1.2,1100::102
dual-59794d75c6-49nkp   vm-003   11.0.3.2,1100::302
dual-59794d75c6-6pgxc   vm-003   11.0.3.3,1100::303
dual-59794d75c6-8sbgj   vm-002   11.0.2.2,1100::202
dual-59794d75c6-tmql4   vm-001   11.0.1.3,1100::103
ipv4-559fc86c45-2lsnv   vm-002   11.0.2.3
ipv4-559fc86c45-6966t   vm-001   11.0.1.4
ipv4-559fc86c45-dt77r   vm-003   11.0.3.4
ipv4-559fc86c45-x2hnx   vm-002   11.0.2.4
ipv4-559fc86c45-zp5lj   vm-001   11.0.1.5
ipv6-6c84b8dd79-556tz   vm-002   1100::204
ipv6-6c84b8dd79-ctznx   vm-001   1100::104
ipv6-6c84b8dd79-jjkr4   vm-002   1100::203
ipv6-6c84b8dd79-l888f   vm-003   1100::304
ipv6-6c84b8dd79-w8jz9   vm-003   1100::305
```
Just works :smile: K8s is pretty cool.

**Comment by uablrek** [comment]
Experiments with services on K8s `v1.24.1`. All PODs above have the same selector ("mserver"). Services are created with;
```yaml
apiVersion: v1
kind: Service
metadata:
  name: mserver-ipv4
spec:
  ipFamilies:
    - IPv4
  selector:
    app: mserver
  ports:
  - port: 5001
    name: mconnect
  - port: 5003
    name: ctraffic
  type: LoadBalancer
---
apiVersion: v1
kind: Service
metadata:
  name: mserver-ipv6
spec:
  ipFamilies:
    - IPv6
  selector:
    app: mserver
  ports:
  - port: 5001
    name: mconnect
  - port: 5003
    name: ctraffic
  type: LoadBalancer
---
apiVersion: v1
kind: Service
metadata:
  name: mserver-dual
spec:
  ipFamilyPolicy: RequireDualStack
  selector:
    app: mserver
  ports:
  - port: 5001
    name: mconnect
  - port: 5003
    name: ctraffic
  type: LoadBalancer
```
Everything works exactly as I hoped. Endpoints and endpointslices are as expected. Access via an IPv4 address reaches all PODs with an IPv4 address (dual-stack and ipv4-only) and access via an IPv6 address reaches all PODs with an IPv6 address (dual-stack and ipv6-only). Below is an example of access from an external machine;
```
vm-201 ~ # kubectl get svc
NAME           TYPE           CLUSTER-IP       EXTERNAL-IP                 PORT(S)                         AGE
kubernetes     ClusterIP      12.0.0.1         <none>                      443/TCP                         9m11s
mserver-dual   LoadBalancer   12.0.19.20       10.0.0.0,1000::1:10.0.0.0   5001:32714/TCP,5003:31646/TCP   8m40s
mserver-ipv4   LoadBalancer   12.0.16.205      10.0.0.1                    5001:31505/TCP,5003:30382/TCP   8m40s
mserver-ipv6   LoadBalancer   fd00:4000::15a   1000::1:10.0.0.2            5001:31633/TCP,5003:32655/TCP   8m40s
vm-201 ~ # mconnect -address 10.0.0.0:5001 -nconn 100
Failed connects; 0
Failed reads; 0
ipv4-746ddb9b9d-2s6z2 10
ipv4-746ddb9b9d-jhhz2 10
dual-746ddb9b9d-b4t8g 10
ipv4-746ddb9b9d-knhhj 10
dual-746ddb9b9d-sl4pb 10
ipv4-746ddb9b9d-gfpcw 9
ipv4-746ddb9b9d-5h5fm 10
dual-746ddb9b9d-brnjj 10
dual-746ddb9b9d-dwckx 11
dual-746ddb9b9d-gfqz9 10
vm-201 ~ # mconnect -address [1000::1:10.0.0.0]:5001 -nconn 100
Failed connects; 0
Failed reads; 0
dual-746ddb9b9d-dwckx 10
dual-746ddb9b9d-sl4pb 10
ipv6-746ddb9b9d-nqpfz 10
ipv6-746ddb9b9d-fjjvl 10
dual-746ddb9b9d-gfqz9 10
ipv6-746ddb9b9d-mr7pt 10
ipv6-746ddb9b9d-7rw7n 10
dual-746ddb9b9d-b4t8g 10
dual-746ddb9b9d-brnjj 10
ipv6-746ddb9b9d-sbd8v 10
```

**Comment by caseydavenport** [comment]
> Soft-standardize one or more annotations and recommend net plugins respect them
> Add a pod API to specify this and recommend net plugins respect them

Speaking for Calico, it would be pretty trivial for us to implement either of these options should this be the direction we go.

Feels like the main use-case that our users would care about is the "sparse IPv4 address space" scenario, where the default would be all pods receive a v6 address but a small subset get marked in some way to also receive a v4 address.

I think this is probably something that with enough time Calico would implement on its own without any standardization on Kubernetes' part, but if we can get buy in from enough implementors then obviously a standard UX would be desirable.

**Comment by thockin** [comment]
So for the record, my thinking:

I think the "v6 everywhere, v4 when specifically requested" case is real and compelling.  It brings an interesting IPAM challenge, though.  If the goal is to economize v4 addresses, you don't want to pre-allocate IP ranges to every node.  So when a pod which DOES want v4 lands on a node, you can't use the simple IPAM mode.  This is a problem for the implementation to solve, it doesn't necessarily mean we should not pursue options here.

As for how to model this API, I am not sure.  We need a way to configure defaults.  I know there's a renewed interest in multi-network and I am cautiously open to that disucssion.  This topic overlaps somewhat, so we should try to be thoughtful about it.

When it comes to how we would scaffold this - I like the idea that it could become parameters to the net plugins, rather than purely the plugins' responsibility, but it's not clear to me (I'm not so steeped in CNI any more :) exactly how that would look.

So I am going to triage-accept this.  It's not a commitment to implement, but acknowledgment that it's not dead-on-arrival.  Like any idea of this scope, it would need a champion to drive it.

/triage accepted

**Comment by uablrek** [comment]
>  If the goal is to economize v4 addresses, you don't want to pre-allocate IP ranges to every node.

I would agree if the ipv4 addresses were Globally Unique (GUA), but I don't think that's the main goal. Global ipv4 addresses are for services (or external load-balancers). As opposed to ipv6 where the intention seem to be to assign GUA/ipv6 to all PODs (and skip NAT) from what I heard on several presentations on KubeCon.

I think a more common scenario is that some legacy application simply can't use ipv6 and is deployed in a _very large_ cluster. In that case it might be OK to pre-allocate (or hand out on demand) IP IPv4 ranges to every node, but rather small ones.

**Comment by caseydavenport** [comment]
> So when a pod which DOES want v4 lands on a node, you can't use the simple IPAM mode. This is a problem for the implementation to solve, it doesn't necessarily mean we should not pursue options here.

Yeah, this use-case poses some problems for something like host-local (unless maybe you only give out a very small v4 CIDR per node). Not every IPAM model will be able to provide for this to the same degree.

**Comment by danwinship** [comment]
> As for how to model this API, I am not sure. We need a way to configure defaults.

There is no obvious place in the core kubernetes configuration to configure default Pod IP family behavior, but that doesn't seem like a problem since you don't configure _available_ Pod IP families in the core kubernetes configuration either. We can just delegate this to the network plugins. The plugin, based on its own configuration, will decide what the default behavior is for assigning IPs to pods, and any pods that don't request otherwise will get that default behavior. (I mean, that's exactly what we have now, except for the implication that there's a way for pods to request otherwise.)

In that case, copying `IPFamilyPolicy`/`IPFamilies` from Services over to Pods might not be quite right, because those fields get defaulted in Services if you leave them unset, but we wouldn't be able to default the corresponding Pod field, because the apiserver won't know what the network plugin's default is.

So maybe, `IPFamilyOverride: [ nil | IPv4Only | IPv6Only | DualStack ]` ? If you don't specify it, it just stays `nil`, meaning "the pod isn't trying to override the plugin's IP family policy".

Kubelet can convert that into CNI `args` or whatever to pass to the plugin. And then we need to figure out what kubelet should do if the plugin ignores the request. (Maybe there could be separate `PreferIPv4Only` and `RequireIPv4Only` etc?)

**Comment by danwinship** [comment]
> > If the goal is to economize v4 addresses, you don't want to pre-allocate IP ranges to every node.
> 
> I would agree if the ipv4 addresses were Globally Unique (GUA), but I don't think that's the main goal.

There are people who are even running out of private-use IPv4 IPs. (I've heard this explicitly, but, eg, it's also implied by https://github.com/kubernetes/kubernetes/issues/110456.)

**Comment by khenidak** [comment]
We need to separate the two problems `pod ip range exhaustion` and `per pod ip family control` the second one is the result of the first one. We should support adding cidr ranges to node cidrs (for CNIs that use them) similar to the work we are doing for services. But beyond that i really don't think we should create a cluster shape where pod can run with single, dual, family specific IPs in dual stack clusters. Not saying it is not a valid requirements, I am saying users can solve for by choosing which ip family they want to listen to in their pods.

**Comment by shaneutt** [comment]
In the SIG Network meeting today we discussed this issue at some length and felt that this is something community members are generally in favor of, but we're uncertain on the "when" in lieu of someone strongly championing it to push it forward. This doesn't really meet the [criteria for help-wanted](https://github.com/kubernetes/community/blob/master/contributors/guide/help-wanted.md#help-wanted) I believe in that I don't expect this has a low barrier to entry, or is a particularly clear task. And partially due to that, a KEP would likely be helpful here so we can reach consensus about _how_ to accomplish it.

So if some interested party would signal willingness to champion it by starting on a KEP, that would seem to be a great next step. Let us know if you need any help in doing so. Otherwise I think it may be realistic to expect that we're going to play lifecycle stale tennis with this one for some time.

**Comment by uablrek** [comment]
Please do not miss https://github.com/kubernetes/kubernetes/issues/109814#issuecomment-1151352002.

> Speaking for Calico, it would be pretty trivial ...

I recommend users what wants this function to request it from Calico first.

**Comment by uablrek** [comment]
An alternative may be to limit ipv4 addresses to some _nodes_ (rather than PODs). I have tried this by using ranges in annotations in the node objects, but the K8s ipam allocation configured by `--allocate-node-cidrs=true` to the kube-controller-manager can perhaps be altered to hand out ipv4 slices to just _some_ nodes.

**Comment by aojea** [comment]
> `-allocate-node-cidrs=true` to the kube-controller-manager can perhaps be altered to hand out ipv4 slices to just _some_ nodes.

it is in alpha https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2593-multiple-cluster-cidrs :)

**Comment by uablrek** [comment]
Cool! How have missed it?

I have configured the cluster, but how do I create a ClusterCIDR object?

I have grep'ed and made a manifest by guessing and tried "kubectl create -f ...", but it doesn't work.

**Comment by aojea** [comment]
@uablrek , you have to enable the feature and the api group, check out latest @sarveshr7 PR , it has instructions on how to use it in kind https://github.com/kubernetes/enhancements/pull/3839/

**Comment by uablrek** [comment]
My problem might be kubectl;
```
k8s-multi-cidr > kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.27.0-alpha.1.228+0e3818e02760af
Kustomize Version: v4.5.7
Server Version: v1.27.0-alpha.1.228+0e3818e02760af
k8s-multi-cidr > kubectl get cc -A
error: the server doesn't have a resource type "cc"
```
Both K8s and kubectl are built from "master" around Sat Feb 4 01:00:12 PM CET.

**Comment by uablrek** [comment]
Um, what is the "api group"? I missed that, and I don't know that it is.

**Comment by uablrek** [comment]
Both `kube-apiserver` and `kube-controller-manager` are started with `--feature-gates MultiCIDRRangeAllocator=true`

**Comment by aojea** [comment]
--runtime-config=networking.k8s.io/v1alpha1=true

https://kubernetes.io/docs/tasks/administer-cluster/enable-disable-api/

**Comment by uablrek** [comment]
Thanks!

**Comment by uablrek** [comment]
Reading about the [Calico ipam confguration](https://projectcalico.docs.tigera.io/reference/cni-plugin/configuration) I think **this problem is already solved for Calico users**.

The MultiCIDRRangeAllocator in K8s allows to limit IPv4 addresses to some nodes, e.g:
```yaml
apiVersion: networking.k8s.io/v1alpha1
kind: ClusterCIDR
metadata:
  name: ipv4
spec:
  nodeSelector:
    nodeSelectorTerms:
      - matchExpressions:
        - {key: wantIPv4, operator: In, values: ["true"]}
  perNodeHostBits: 7
  ipv4: "172.20.0.0/24"
  ipv6: "fd00:1000::/120"
---
apiVersion: networking.k8s.io/v1alpha1
kind: ClusterCIDR
metadata:
  name: default
spec:
  perNodeHostBits: 16
  ipv6: "fc00:2000::/96"
```

But that requires support in the CNI-plugin. Flannel doesn't work (stuck in ContainerCreating), but Antrea does :smiley: 
```
# kubectl get pods -o 'custom-columns=NAME:metadata.name,NODE:spec.node
Name,IPs:status.podIPs[*].ip'
NAME                      NODE     IPs
alpine-65f85598fb-488l7   vm-004   fc00:2000::1:2
alpine-65f85598fb-5pn62   vm-004   fc00:2000::1:5
alpine-65f85598fb-6fzsh   vm-002   172.20.0.4,fd00:1000::4
alpine-65f85598fb-6rv7x   vm-002   172.20.0.5,fd00:1000::5
alpine-65f85598fb-9whx8   vm-004   fc00:2000::1:4
alpine-65f85598fb-b9xf7   vm-004   fc00:2000::1:3
alpine-65f85598fb-bhr8p   vm-002   172.20.0.2,fd00:1000::2
alpine-65f85598fb-cbrz6   vm-003   172.20.0.132,fd00:1000::84
alpine-65f85598fb-fd6bm   vm-002   172.20.0.3,fd00:1000::3
alpine-65f85598fb-h65mc   vm-005   fc00:2000::2:3
alpine-65f85598fb-jwvcl   vm-003   172.20.0.133,fd00:1000::85
alpine-65f85598fb-mqqrs   vm-003   172.20.0.131,fd00:1000::83
alpine-65f85598fb-p5rv6   vm-005   fc00:2000::2:4
alpine-65f85598fb-q56jj   vm-005   fc00:2000::2:2
alpine-65f85598fb-q8vzh   vm-005   fc00:2000::2:5
alpine-65f85598fb-vxt4d   vm-003   172.20.0.130,fd00:1000::82
```

**Comment by aojea** [comment]
@manuelbuil did you get to implement this in flannel?

**Comment by uablrek** [comment]
For the record, I used `flannel:v0.19.0`

**Comment by danwinship** [comment]
> Option 3 starts to feel a whole lot like multi-network...

> If you model the "pod network" as 2 networks (v4, v6) and allow pods to specificy which one(s) they want, maybe it can be a fit. Maybe. Very speculative since we don't have multi-network as a built-in concept.

> I think the "v6 everywhere, v4 when specifically requested" case is real and compelling. It brings an interesting IPAM challenge, though. If the goal is to economize v4 addresses, you don't want to pre-allocate IP ranges to every node.

> So then a cluster could have a mix of IPv6-only nodes and dual-stack nodes, and, eg, you could taint the dual-stack nodes, and then pods would only get dual-stack IPs if they were explicitly marked as "tolerating" dual-stack. (Or you could use various other means of correctly scheduling pods to nodes.)

This all does sound _a lot_ like multi-network... You have a default IPv6 pod network that is available everywhere, and an optional IPv4 pod network that is only available on certain nodes, and pods that need IPv4 access would specify that they need to be attached to that network, in the appropriate way.

(cc @mskrocki @MikeZappa87 )

**Comment by uablrek** [comment]
Difference is that this is _the_ K8s network, just different ip-families. No multi-network is needed, is works with an unmodified K8s with a bit "unusual" IPAM. I wrote an IPAM where you can limit assignment of IPv4 addresses to some namespaces, [kube-node](https://github.com/Nordix/ipam-node-annotation#limit-ipv4-address-allocation).

This will also allow a higher `maxPods` than 110 (which seem very low on modern HW), because that comes from [IPv4 limitations](https://arunprakashpj.medium.com/kubernetes-what-if-the-pod-limit-crosses-110-740e2e1a21e3). Please see also https://github.com/kubernetes-sigs/node-ipam-controller/issues/5.

**Comment by thockin** [comment]
xref https://github.com/kubernetes/kubernetes/issues/114702 which is about dual-stack HCNP.

Discussed at sig-net today, with some strong opinions on whether and how we want to support mixed-single-stack clusters.

1) If it was trivial to add this support (it's not, but suppose it was) - do we have philosophical objections?  It would add fuel to the need for dual-stack apiuserver service (`kubernetes.default`).  It would also make the need for dual-stack HCNP more obvious.  It was pointed out that CNIs can already do this and we do not prevent it.  It was also pointed out that just because a pod gets assigned 2 IPs doesn't mean it *works* on both, and so dual-stack HCNP would still be inaccurate in some cases.

2) If it was not trivial to add, but someone felt strongly enough to do the work, would we be OK with that?

3) Does anyone feel strongly enough, or have real customer demand for this, who is willing to take on the work if we were conceptually OK with it?

**Comment by uablrek** [comment]
Well, two exceptions (equally rare):
Single-stack PODs of the "other" (not the default) family that needs to communicate with the API-server

**Comment by danwinship** [comment]
@aojea and I have been arguing about this, with the argument basically coming down to the question of what the goal of dual-stack in Kubernetes is:
1. A dual-stack cluster exists to provide users with full dual-stack connectivity everywhere. Every pod can reach every other pod, just like in a single-stack cluster. Every pod can reach both IPv4-only and IPv6-only servers outside the cluster, and (ideally) both IPv4-only and IPv6-only clients outside the cluster can reach every LB/ingress/gateway into the cluster.
2. A dual-stack cluster exists to provide users with IPv4 connectivity where they want IPv4 connectivity, and IPv6 connectivity where they want IPv6 connectivity (which obviously also includes having dual-stack connectivity where they want want dual-stack connectivity). If you _want_ the behavior described in 1, you're certainly allowed to set things up that way, but you don't _have to_.

Ignoring the question of what we _should_ do, the current implementation leans strongly toward 2, in large part due to the fact that there is no cluster-wide networking configuration and so you can bring up a cluster with single-stack IPv4 pods, single-stack IPv6 nodes, and dual-stack services, though obviously not everything will work in that case. OTOH, the current _documentation_ probably leans more toward 1.

> It was also pointed out that just because a pod gets assigned 2 IPs doesn't mean it works on both, and so dual-stack HCNP would still be inaccurate in some cases.

If a pod chooses to be part of a dual-stack service, but only listens on 1 IP family when it has dual-stack IPs, it's broken and you deserve to lose.

**Comment by sebhoss** [comment]
Calico 3.30 now supports a `cni.projectcalico.org/ipFamilies` annotation to restrict IP families for certain pods ([PR](https://github.com/projectcalico/calico/pull/9544))
